packages/babel-parser/src/parser/node.js

```diff
@@ -39,13 +43,51 @@ class Node implements NodeBase {
         key !== "trailingComments" &&
         key !== "innerComments"
       ) {
-        // $FlowIgnore
         newNode[key] = this[key];
       }
     }
 
     return newNode;
+  };
+}
+
+function clonePlaceholder(node: any): any {
+  return cloneIdentifier(node);
+}
+
+export function cloneIdentifier(node: any): any {
+  // We don't need to clone `typeAnnotations` and `optional`: because
+  // cloneIdentifier is only used in object shorthand and named import/export.
+  // Neither of them allow type annotations after the identifier or optional identifier
+  const { type, start, end, loc, range, extra, name } = node;
+  const cloned = Object.create(NodePrototype);
+  cloned.type = type;
+  cloned.start = start;
+  cloned.end = end;
+  cloned.loc = loc;
+  cloned.range = range;
+  cloned.extra = extra;
+  cloned.name = name;
+  if (type === "Placeholder") {
+    cloned.expectedNode = node.expectedNode;
+  }
+  return cloned;
+}
+
+export function cloneStringLiteral(node: any): any {
+  const { type, start, end, loc, range, extra } = node;
+  if (type === "Placeholder") {
+    return clonePlaceholder(node);
   }
+  const cloned = Object.create(NodePrototype);
```

Comment:
'Object.create' is so much slower than a constructor. And overall it's slow. 

https://stackoverflow.com/questions/43372746/why-so-bad-the-performance-of-the-nodejs-object-creation-if-num-of-props-greater

Reply:
We will remove `Object.create` here in Babel 8.

Reply:
Can't you just use the '__proto__' trick?

Reply:
`__proto__` is slower than `Object.create`:

```
baseline 256 length-1 named export: 5_650 ops/sec ±45.38% (0.177ms)
baseline 512 length-1 named export: 4_129 ops/sec ±1.16% (0.242ms)
baseline 1024 length-1 named export: 1_977 ops/sec ±1.31% (0.506ms)
baseline 2048 length-1 named export: 894 ops/sec ±2.1% (1.118ms)
current 256 length-1 named export: 5_269 ops/sec ±29.84% (0.19ms)
current 512 length-1 named export: 3_003 ops/sec ±2.24% (0.333ms)
current 1024 length-1 named export: 1_473 ops/sec ±2.04% (0.679ms)
current 2048 length-1 named export: 710 ops/sec ±1.6% (1.409ms)
```

Reply:
> create a new obj based on the given params again

I agree. But we should get rid of `loc` and `ranges` first so the parameters can be limited to `start`, `end`, `value` and `type` (solely for placeholder support).

Reply:
or just pack the loc data in a series of binary numbers

Reply:
Well there are also `filename` and `identifierName` in the `loc`.